shiny/_main_create.py

```diff
@@ -121,13 +264,124 @@ def use_template_internal(
 
         webbrowser.open(url)
         sys.exit(0)
-    elif template == "js-component":
-        js_component_questions(dest_dir=dest_dir, package_name=package_name)
-        return
-    elif template in package_template_choices.values():
-        js_component_questions(template, dest_dir=dest_dir, package_name=package_name)
+    elif question_state == "_js-component":
+        use_internal_package_template(dest_dir=dest_dir, package_name=package_name)
+    elif question_state == "_chat-ai":
+        use_internal_chat_ai_template(dest_dir=dest_dir, package_name=package_name)
     else:
-        app_template_questions(template, mode, dest_dir=dest_dir)
+        valid_choices = [t.name for t in app_templates + pkg_templates]
+        if question_state not in valid_choices:
+            raise click.BadOptionUsage(
+                "--template",
+                f"Invalid value for '--template' / '-t': {question_state} is not one of "
+                + f"""'{"', '".join(valid_choices)}'.""",
+            )
+
+
+def use_internal_package_template(
+    dest_dir: Optional[Path] = None,
+    package_name: Optional[str] = None,
+):
+    input = questionary.select(
+        "What kind of component do you want to build?",
+        choices=[
+            *choice_from_templates(shiny_internal_templates.packages),
+            back_choice,
+            cancel_choice,
+        ],
+        style=styles_for_questions,
+    ).ask()
+
+    if input == "back":
+        use_internal_template()
+        return
+
+    if input is None or input == "cancel":
+        sys.exit(1)
+
+    template = template_by_name(shiny_internal_templates.packages, input)
+
+    if template is None:
+        # This should be valid because we're selecting from the list of templates
+        # but just in case and to make type checkers happy
+        raise ValueError(f"Package template for {input} not found.")
+
+    package_template_questions(template, dest_dir=dest_dir, package_name=package_name)
+
+
+def use_internal_chat_ai_template(
+    input: str | None = None,
+    dest_dir: Optional[Path] = None,
+    package_name: Optional[str] = None,
+):
+    if input is None:
+        input = questionary.select(
+            "Which kind of generative AI template would you like to use?",
+            choices=[
+                Choice(title="By provider...", value="_chat-ai_hello-providers"),
+                Choice(title="Enterprise providers...", value="_chat-ai_enterprise"),
+                back_choice,
+                cancel_choice,
+            ],
+            style=styles_for_questions,
+        ).ask()
```

Comment:
It doesn't feel great to be "exporting" `_templates.json` with no way to express choice groups (i.e., sub-menus).

If there was one `_templates.json` per `shiny create`, it'd be fairly easy to support choice groups, but with one `_templates.json` per app, that seems harder

Reply:
It's so much easier to work with a single json file per template than in one large JSON file. With atomic template metadata, it's much easier to keep track of the metadata for the template. For example, I just moved the chat templates from `examples/chat` to `shiny/templates/chat` and had to change one string in our code base. Then I renamed `shiny/templates/app-templates` to `shiny/template/app` and again made one small change.

With the `--github` flag, it's easy to do one-level choice groups with the current setup. Imagine we put the chat templates in `py-shiny-templates`. Then you can construct a `shiny create` command that only lists those chat templates, ignoring everything else in the repo:

```bash
shiny create --github posit-dev/py-shiny-templates:chat/hello-providers
```

We could certainly implement choice groups using an additional metadata file, but organizing templates by directory with atomic template files is pretty powerful. In practice, I think we'd use the GitHub repo path formulation to limit choices for a certain type of template within a template repo.

Overall, this approach feels like a good 80% of what we want without rewriting from scratch, over investing in abstractions, or closing off the opportunity to build on this in the future.

I guess the big thing is that, if we find we have a need we can easily extend the current approach to build submenus. I certainly see the value in having support for submenus, but I don't think the need is strong enough at this moment for it to be needed.

Reply:
> easily extend the current approach to build submenus

If you have something in mind, would you mind writing it up in an issue?

Reply:
https://github.com/posit-dev/py-shiny/issues/1647